When damage is auto-rolled, roll crit damage when appropriate

ian612 opened the pull request:
This adds a mechanism to automatically roll critical damage rather than normal when the auto-roll damage option is selected and a critical hit is rolled. This also adds `isCriticalSuccess()` and `isFumble()` helper methods to `DiceSFRPG` to assist with this.

Resolves #1491

danimrath commented:
If you do not have auto quick roll turned on and attack with or without shift click it will pop a prompt for the damage but clicking the normal button instead of critical still does a crit. 

It feels like this would also be good for shift click on the damage button of an attack card to roll crit damage.

ian612 commented:
I've fixed the bug you mentioned; clicking the "Normal" or "Critical" damage buttons in the damage dialog now rolls the selected damage variation regardless of the attack roll result. If _quick roll_ and _roll damage with attack_ are both selected, it rolls no damage on a fumble, critical damage on a critical hit, and normal damage on any other value.

In addition, I've added functionality for rolling critical damage directly from the character sheet, but using the control key instead of shift, since Shift is used to toggle between auto-roll and dialog box display already.

danimrath commented:
The only things left I have to say on it are that 
1) If (d.total >= (d.options.critical || critValue))  feels off. Feels like it should be a one or the other not an OR situation and I'm not sure >= is appropriate? I'm wondering what your intention is here as it does technically work but just feels like this opens the door to potential unintended effects down the line.
2) It would be great if the damage button on the attack card in chat interacted with this behavior as well. Not sure if that is possible or not but basically if you shift click or auto roll from the chat card having it roll crit appropriately.

ian612 commented:
Yeah you're right, looking at it again I'm realizing it's not doing quite what I wanted. What I want it to do is default to just checking against the critical value stored in the `roll` object, and only check against the `critValue` variable if it's passed in explicitly (allowing the method to be a bit more versatile). I've updated to make that change.

I used `>=`/`<=` on the methods as this makes them compatible with cases where a crit happens on a larger range of numbers, such as a 19-20. I don't think there's anything in the system right now to take advantage of that, but the bones are there for it at least.

Regarding point 2, I think it actually does work the way you're suggesting (assuming I'm understanding correctly). When you shift-click those chat card attack/damage buttons, they skip the dialog and roll. If you Shift+Ctrl+click on the damage button, it rolls crit damage. If the quick roll feature is turned on, Ctrl+click on the damage button rolls crit damage, as expected.

ian612 commented on src/module/item/item.js:
```diff
@@ -1351,6 +1351,7 @@ export class ItemSFRPG extends Mix(foundry.documents.Item).with(ItemActivationMi
         return DiceSFRPG.damageRoll({
             event: event,
             parts: parts,
+            linkedAttackRoll: options.linkedAttackRoll ?? false,
```
I hadn't considered that but I actually like that more than how I've done it, with the additional field. I might try it out and see if it gives me any issues.

ian612 replied:
I gave it a try locally and it works, but after some consideration I think I'll leave it as is. The purpose of the roll context objects are to look values up via `@` tags, so it probably isn't the cleanest to try to also use them to hold optional roll data.
